### PR TITLE
Support custom projects

### PR DESCRIPTION
**Pull request by sebastienros**
I still need to update the repos.targets file to reflect the changes (definition files) and/or the CI job.

### CODE REVIEW

File: src/Benchmarks.ServerJob/ServerJob.cs [hunk 1]
```diff
@@ -29,31 +28,21 @@ public class ServerJob : IIdentifiable
...
-                return _sources ?? Enumerable.Empty<Source>();
-            }
-            set
-            {
-                _sources = value;
-            }
-        }
+        public IList<Source> Sources { get; set; } = new List<Source>();
+
+        public Source Source { get; set; } = new Source();
```
Comment (mikeharder) [inline src/Benchmarks.ServerJob/ServerJob.cs]: Rename either `Sources` or `Source` to clarify the difference, or at least add a comment.

File: src/BenchmarksClient/Startup.cs [hunk 2]
```diff
@@ -158,6 +170,48 @@ public static int Main(string[] args)
...
+
+            using (var response = _httpClient.SendAsync(CreateHttpMessage(job)).GetAwaiter().GetResult())
+            {
+                var responseContent = response.Content.ReadAsStringAsync().GetAwaiter().GetResult();
+                job.LatencyFirstRequest = stopwatch.Elapsed;
+            }
+
+            Log("Measuring single connection latency");
+
+            // This could be done during the Client job but we are already measuring the Startup time here.
```
Comment (mikeharder) [inline src/BenchmarksClient/Startup.cs]: Outdated comment

File: src/Benchmarks.ServerJob/ServerJob.cs [hunk 3]
```diff
@@ -29,31 +28,21 @@ public class ServerJob : IIdentifiable
...
+        public string Arguments { get; set; }
```
Comment (mikeharder) [inline src/Benchmarks.ServerJob/ServerJob.cs]: Unused property?

File: src/BenchmarksDriver/Headers.cs [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+namespace BenchmarksDriver
+{    
+    public enum Headers
+    {
+        None,
+        Html,
+        Plaintext,
+        Json
+    }
+}
```
Comment (mikeharder) [inline src/BenchmarksDriver/Headers.cs]: Add trailing newline

File: src/BenchmarksDriver/JobOptions.cs [hunk 5]
```diff
@@ -0,0 +1,9 @@
+using Newtonsoft.Json;
+
+namespace BenchmarksDriver
+{    
+    public class JobOptions
+    {
+        public string[] Paths { get; set; }
+    }
+}
```
Comment (mikeharder) [inline src/BenchmarksDriver/JobOptions.cs]: Add trailing newline

File: src/BenchmarksDriver/Program.cs [hunk 6]
```diff
@@ -160,28 +77,44 @@ public static int Main(string[] args)
...
+            var argumentsOption = app.Option("--arguments",
+                "Arguments to pass to the application. (e.g., \"--raw true\")", CommandOptionType.SingleValue);
+            var portOption = app.Option("--port",
+                "The port used to request the benchmarked application. Default is 5000.", CommandOptionType.SingleValue);
+            var repositoryOption = app.Option("-r|--repository",
+                "Git repository containing the project to test.", CommandOptionType.SingleValue);
+            var projectOption = app.Option("--projectFile",
+                "Relative path of the project to test in the repository. (e.g., \"src/Benchmarks/Benchmarks.csproj)\"", CommandOptionType.SingleValue);
+            var branchOrCommitOption = app.Option("-b|--branchOrCommit",
+                "Branch name of commit hash to checkout.", CommandOptionType.SingleValue);
```
Comment (mikeharder) [inline src/BenchmarksDriver/Program.cs]: Typo `of` -> `or`

File: src/BenchmarksDriver/Program.cs [hunk 7]
```diff
@@ -215,67 +148,129 @@ public static int Main(string[] args)
...
+                    if (!jobOptions.Paths.Any(p => string.Equals(p, serverJob.Path, StringComparison.OrdinalIgnoreCase)) &&
+                        !jobOptions.Paths.Any(p => string.Equals(p, "/" + serverJob.Path, StringComparison.OrdinalIgnoreCase)))
+                    {
+                        Console.WriteLine($"Scenario '{scenarioName}' does not support {pathOption.LongName} '{pathOption.Value()}'. Choose from:");
+                        Console.WriteLine($"'{string.Join("', '", jobOptions.Paths)}'");
+                        return 6;
+                    }
+                }
+                
+                // These properties are mandatory on the command line and can't be set in the job definitions
```
Comment (mikeharder) [inline src/BenchmarksDriver/Program.cs]: Remove "mandatory"

File: src/BenchmarksServer/BenchmarksServer.csproj [hunk 8]
```diff
@@ -1,4 +1,4 @@
-<Project Sdk="Microsoft.NET.Sdk.Web">
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (mikeharder) [inline src/BenchmarksServer/BenchmarksServer.csproj]: Why stop using the web SDK?

File: src/BenchmarksServer/Startup.cs [hunk 9]
```diff
@@ -313,27 +321,27 @@ void CleanJob()
...
         {
             // It's possible that the user specified a custom branch/commit for the benchmarks repo,
             // so we need to add that to the set of sources to restore if it's not already there.
             //
             // Note that this is also going to de-dupe the repos if the same one was specified twice at
             // the command-line (last first to support overrides).
-            var repos = new HashSet<Source>(job.Sources, SourceRepoComparer.Instance);
+            var repos = new HashSet<Source>(job.ReferenceSources, SourceRepoComparer.Instance);
 
             // This will no-op if 'benchmarks' was specified by the user.
```
Comment (mikeharder) [inline src/BenchmarksServer/Startup.cs]: Remove comment which is no longer correct.

File: src/BenchmarksServer/Startup.cs [hunk 10]
```diff
@@ -360,85 +368,83 @@ private static string CloneRestoreAndBuild(string path, ServerJob job, string do
...
+            var buildToolsPath = Path.Combine(path, "buildtools");
+            
+            await DownloadBuildTools(buildToolsPath);
+
+            Log.WriteLine("Installing dotnet runtimes and sdk");
+
+            // Computes the location of the benchmarked app
+            var benchmarkedApp = Path.Combine(path, benchmarkedDir, Path.GetDirectoryName(job.Source.Project));
+
+            // Degines SDK will be installed. Using "" downloads the latest SDK.
```
Comment (mikeharder) [inline src/BenchmarksServer/Startup.cs]: Typo `Degines`?

File: src/BenchmarksServer/Startup.cs [hunk 11]
```diff
@@ -360,85 +368,83 @@ private static string CloneRestoreAndBuild(string path, ServerJob job, string do
...
             // Build and Restore
-            var benchmarksApp = Path.Combine(benchmarksRoot, "src", "Benchmarks");
             var dotnetExecutable = GetDotNetExecutable(dotnetHome);
 
             // Project versions must be higher than package versions to resolve those dependencies to project ones as expected.
             // Passing VersionSuffix to restore will have it append that to the version of restored projects, making them
             // higher than packages references by the same name.
-            ProcessUtil.Run(dotnetExecutable, "restore /p:VersionSuffix=zzzzz-99999", workingDirectory: benchmarksApp, environmentVariables: env);
-            ProcessUtil.Run(dotnetExecutable, $"build -c Release", workingDirectory: benchmarksApp, environmentVariables: env);
+            var buildParameters = $"/p:BenchmarksAspNetCoreVersion={job.AspNetCoreVersion} /p:BenchmarksNETStandardImplicitPackageVersion={job.AspNetCoreVersion} /p:BenchmarksNETCoreAppImplicitPackageVersion={job.AspNetCoreVersion} /p:BenchmarksRuntimeFrameworkVersion=2.0.0";
```
Comment (mikeharder) [inline src/BenchmarksServer/Startup.cs]: Split into shorter lines

File: src/BenchmarksServer/BenchmarksServer.csproj (hunk 8 above)
Comment (sebastienros) [reply src/BenchmarksServer/BenchmarksServer.csproj]: Couldn't start the debugger from VS as it's a console app actually

**Pull request merged**